Temporal Nexus: Main Encyclopedia (#3053)
temporalio/documentation: docs/encyclopedia/temporal-nexus-overview.mdx

---
id: nexus-overview
title: Temporal Nexus Overview
sidebar_label: Overview
description: Temporal Nexus is designed to connect durable executions across team and Namespace boundaries with an integrated Temporal SDK experience, promoting a modular architecture for sharing your team’s capabilities via a well-defined Nexus Service API contract.
toc_max_heading_level: 4
slug: /temporalnexus/overview
keywords:
  - modular architecture
  - Namespace boundaries
  - durable execution
tags:
  - distributed systems
  - integration
  - temporal
---

> 

Author: Most of this content here is the same content from the [Evaluate page](https://temporal-documentation-git-feature-landing-page-nexus.preview.thundergun.io/nexus). It's going to be repetitive and users will end up reading it twice. Ideally, users will start at the evaluate page then go to the encyclopedia next. Is there a way to sumarize the first four paragraphs. Otherwise, we could keep the first paragraph as the 2-liner intro about Nexus and jump into the Nexus RPC and continue from there.

## Core Components {#nexus-core-components}

> Nexus is designed to connect durable executions across team and Namespace boundaries with an integrated Temporal SDK experience, built-in Nexus Machinery that processes Nexus RPC calls on your behalf, and promotes a more modular architecture for sharing a subset of your team’s capabilities via a well-defined Nexus Service API contract.

Reviewer: ```suggestion
Nexus is designed to connect durable executions across team and Namespace boundaries with an integrated Temporal SDK experience. It provides built-in Nexus Machinery that processes Nexus RPC calls on your behalf, and promotes a more modular architecture for sharing a subset of your team’s capabilities via a well-defined Nexus Service API contract.
```
This sentence is long. Let's split it.

Author: agreed.

> Nexus Services are exposed from a Nexus Endpoint that is created in the Nexus API Registry, along with service documentation that enable others to use the Nexus Services provided by a Nexus Endpoint. This enables teams to compose Nexus Services into their Workflows, both within and across Namespace boundaries.

Reviewer: ```suggestion
Nexus Services are exposed from a Nexus Endpoint that is created in the Nexus API Registry, along with service documentation that enable others to use the Nexus Services provided by a Nexus Endpoint. 
This enables teams to compose Nexus Services into their Workflows, both within and across Namespace boundaries.
```

> Temporal has built-in Nexus machinery that performs the low-level Nexus RPC operations on the wire and provides an integrated Temporal SDK experience to create Nexus Services and use them from a caller Workflow via a Nexus Endpoint in the Nexus API Registry.

Author: I took out core components since we have the subpages and may not be entirely neccessary.

**Nexus API Registry:** A `Nexus API Registry` is a list of Nexus `Endpoints` available within a Temporal environment. 
Adding a Nexus `Endpoint` allows Nexus Operation requests to be sent to the Worker listening on the Endpoint’s target Namespace and Task Queue, from a caller Workflow using the name of the Endpoint in the Nexus API Registry.

> **Nexus Endpoints:** A Nexus `Endpoint` is like an API proxy that can serve one or more Nexus `Services` from a `Worker` listening on a target Namespace and Task Queue. 

Reviewer: ```suggestion
**Nexus Endpoints:** A Nexus `Endpoint` is an API proxy that can serve one or more Nexus `Services` from a `Worker` listening on a target Namespace and Task Queue. 
```

This enables presenting a clean API contract and hiding the underlying Temporal implementation, which may consist of internal `Workflows`, `Signals`, and other Temporal primitives – or arbitrary code. 
In Temporal Cloud an endpoint’s access-control policy determines which callers are allowed to access a Nexus `Endpoint`.

**Nexus Services:** A Nexus `Service` is a named collection of arbitrary-duration Nexus `Operations` that provide an API contract suitable for sharing across team boundaries. 
Nexus Services are registered with a `Worker` that is listening on the target Namespace and Task Queue for an `Endpoint`.

**Nexus Operations:** A Nexus `Operation` is not intrinsically durable, but can be made durable by backing it with Temporal primitives like a `Workflow`, `Update`, or `Signal`.
The Temporal SDK provides helper functions to create Nexus Operations that may be registered with a Nexus Service in a `Worker`. 

> Synchronous Nexus `Operations` can also execute arbitrary code, but unlike `Activities` they have a short request timeout (under 10 seconds) and don’t support [heartbeats](https://community.temporal.io/t/best-practices-for-long-running-activities/934).

Reviewer: Should we pull this community article into docs instead of linking to it? Feels odd linking to a community forum.

Author: Taking out the core components section for now since there's already sections dedicted for each one.

> Nexus RPC is designed for arbitrary-duration operations whose lifetime may extend beyond a traditional RPC and is an underpinning to connect durable executions within and across Namespaces, clusters and regions – with a clean API contract to streamline multi-team collaboration. Any service can be exposed via a Nexus endpoint as a set of sync or async Nexus operations – the latter provides an operation identity and a uniform interface to get the status of an operation or its result, receive a completion callback, or cancel the operation.

Reviewer: Personally, the em dashes make the sentences too long. I'd rather us remove them and make the following part proper sentences.

> Nexus RPC is designed for arbitrary-duration operations whose lifetime may extend beyond a traditional RPC and is an underpinning to connect durable executions within and across Namespaces, clusters and regions – with a clean API contract to streamline multi-team collaboration. Any service can be exposed via a Nexus endpoint as a set of sync or async Nexus operations – the latter provides an operation identity and a uniform interface to get the status of an operation or its result, receive a completion callback, or cancel the operation.

Reviewer: ```suggestion
Nexus RPC is designed for arbitrary-duration operations whose lifetime may extend beyond a traditional RPC and is an underpinning to connect durable executions within and across Namespaces, clusters and regions – with a clean API contract to streamline multi-team collaboration. 
Any service can be exposed via a Nexus endpoint as a set of sync or async Nexus operations – the latter provides an operation identity and a uniform interface to get the status of an operation or its result, receive a completion callback, or cancel the operation.
```

> Nexus Services and Operations are often registered in the same Worker as the underlying Temporal primitives they abstract, and use the same task-queue worker polling architecture as Workflows and Activities.

Reviewer: ```suggestion
Nexus Services and Operations are often registered in the same Worker as the underlying Temporal primitives they abstract, and use the same Task Queue worker polling architecture as Workflows and Activities.
```

> Nexus Services and Operations are often registered in the same Worker as the underlying Temporal primitives they abstract, and use the same task-queue worker polling architecture as Workflows and Activities.

Author: This is the overview, should it be removed? @prasek

> To make a Nexus Service available for others to use, a Nexus Endpoint must be created in the Nexus API Registry with a target Namespace and task-queue. 

Reviewer: ```suggestion
To make a Nexus Service available for others to use, a Nexus Endpoint must be created in the Nexus API Registry with a target Namespace and Task Queue.
```
Being consistent with the rest of the docs.

> In pre-release, all Nexus requests will be sent to the single target Namespace and task-queue for the Endpoint. Future releases may support multiple Endpoint targets, similar to how an API proxy can have multiple targets.

Reviewer: ```suggestion
In pre-release, all Nexus requests will be sent to the single target Namespace and Task Queue for the Endpoint. Future releases may support multiple Endpoint targets, similar to how an API proxy can have multiple targets.
```

Author: this overview was replaced

> In pre-release, all Nexus requests will be sent to the single target Namespace and task-queue for the Endpoint. Future releases may support multiple Endpoint targets, similar to how an API proxy can have multiple targets.

Reviewer 2: I would just drop this for now, I don't think it provides value here. (And by the time we publish this, we'll be in public preview).

When you start a Nexus Operation in a caller Workflow, a command is sent to Temporal to schedule the Operation, and then the Temporal Nexus Machinery is responsible for making the Nexus RPC calls on your behalf. 
This means you don’t have to use Nexus RPC directly, only the Temporal SDK along with a Temporal Service.

> Similar to executing an Activity, for a Nexus Operation:

Reviewer 2: I'd just clarify this isn't something that the average user need to know about.

```suggestion
Similar to executing an Activity, for a Nexus Operation, Temporal runs the following flow behind the scenes:
```

1. Caller Workflow will schedule a Nexus Operation using the Temporal SDK, which sends a `ScheduleNexusOperation` command in the existing `RespondWorkflowTaskCompleted` RPC which is recorded as `NexusOperationScheduled` in the caller’s Workflow history.
2. Nexus handler Worker uses the Temporal SDK to `PollNexusTaskQueue` and process the Nexus request.

> 3. Nexus handler Worker sends a `RespondNexusTaskCompleted` via the SDK, with the result of the Nexus request, which may be one of: Started, Completed, Failed, TimedOut, Canceled,  see Nexus Operation Events for more info

Reviewer 2: Spacing here is a bit weird.

> 4. Caller Workflow then processes the Nexus Operation event via `PollWorkflowTaskQueue`.

Reviewer 2: ```suggestion
4. Temporal records a corresponding Nexus event in the caller Workflow's event history, e.g. `NexusOperationCompleted`.
5. Caller Workflow is notified of progress when it polls for the next Workflow Task via `PollWorkflowTaskQueue`.
```

> See also: [Nexus Architecture on GitHub](https://github.com/temporalio/temporal/blob/main/docs/architecture/nexus.md).

Reviewer 2: I would use a permalink, in case this content moves.

## Key Capabilities

Nexus support is available in Temporal Cloud, OSS Server, and Temporal SDKs:

- Workers that handle Nexus Operations, listening on a Namespace and Task Queue

> - Nexus API registry, to create Nexus Endpoints that forward requests to a target Namespace and Task Queue, so others can use Nexus services via a Nexus endpoint.

Reviewer: ```suggestion
- Nexus API registry, to create Nexus Endpoints that forward requests to a target Namespace and Task Queue, so others can use Nexus services via a Nexus endpoint
```

Reviewer 2: Replace "forwards" with "routes"

Author: The overview has been replaced via internal doc.

> - Start Nexus operations from Temporal Workflows across team & Namespace boundaries

Reviewer 2: ```suggestion
- Start Nexus Operations from Temporal Workflows across team & Namespace boundaries
```

- Observe and troubleshoot end-to-end execution

Temporal Cloud provides enhanced security, observability, and multi-region routing:

- Multi-region routing from Caller to destination Namespace in a Temporal Cloud account
- Control who can manage a Nexus Endpoint with a new Nexus Endpoint Admin permission
- Control who can use a Nexus Endpoint within an account, using Namespace allowlists
- Observe Nexus audit logs and observability metrics via cloud integrations

#### Integrated Temporal SDK experience

- Run Nexus Services in a Temporal Worker:
  - Abstract and expose a subset of Temporal primitives for others to use

>   - Can also execute arbitrary code

Reviewer 2: I'd clarify this a bit. Hard to understand this sentence in this context

>   - Typically the same Worker as the underlying Temporal primitives

Reviewer 2: I don't understand what this sentence is trying to convey.

  - Same queue-based Worker architecture, for Worker fleet load balancing

>   - No separate deployment for a bespoke service wrapper

Reviewer 2: Not sure users will understand what a bespoke service wrapper is.

  - Polling architecture enables workers to run in a broad range of runtime environments

> - Use Nexus Services from a Workflow

Reviewer 2: Is this a separate list? It doesn't seem like it belongs in the list above and doesn't have a colon.

Reviewer 2: Also, seems redundant with the item below.

  - Start a Nexus Operation from a Temporal Workflow

#### Built-in Nexus Machinery

- Built-in support in Temporal Server and Temporal Cloud
- Connects Nexus callers and handlers through Nexus Endpoints
- Routes Nexus calls across team and Namespace boundaries
- Automatic retries
- Integrated into Workflow event history for end-to-end debugging

> - Supports metrics, traces, and logs

Reviewer 2: Can we turn these sections into paragraphs instead of these bullet lists? This doesn't feel very polished to me and lacks context.

Author: Agreed. These were taken out and replaced with a new overview. I do want to try to avoid short bullet point lists.
